roll/websocket.py

```diff
@@ -0,0 +1,124 @@
+
+        self.handler = handler
+        if subprotocols:
+            subprotocols = frozenset(subprotocols)
+        self.subprotocols = subprotocols
+
+    def handshake(self, request):
+        """Websocket handshake, handled by `websockets`
+        """
+        headers = []
+
+        def get_header(k):
+            return request.headers.get(k.upper(), '')
+
+        def set_header(k, v):
+            headers.append((k, v))
+
+        try:
+            key = handshake.check_request(get_header)
+            handshake.build_response(set_header, key)
+        except InvalidHandshake:
+            raise RuntimeError('Invalid websocket request')
+
+        subprotocol = None
+        ws_protocol = get_header('Sec-Websocket-Protocol')
+        if self.subprotocols and ws_protocol:
+            # select a subprotocol
+            client_subprotocols = tuple(
+                (p.strip() for p in ws_protocol.split(',')))
+            for p in client_subprotocols:
+                if p in self.subprotocols:
+                    subprotocol = p
+                    set_header('Sec-Websocket-Protocol', subprotocol)
+                    break
+
+        # write the 101 response back to the client
+        # We will use a real Response in the future.
+        rv = b'HTTP/1.1 101 Switching Protocols\r\n'
+        for k, v in headers:
+            rv += k.encode('utf-8') + b': ' + v.encode('utf-8') + b'\r\n'
+        rv += b'\r\n'
+        request.transport.write(rv)
+
+        # Return the subprotocol agreed upon, if any
+        return subprotocol
+
+    def switch_protocol(self, request):
+        # The websocket handshake agrees on the subprotocol to use.
+        # It then writes a response to the client, before we go any further.
+        subprotocol = self.handshake(request)
+
+        # Creation of the new protocol, with the agreed upon subprotocol
+        websocket = WebSocketCommonProtocol(
+            timeout=self.timeout,
+            max_size=self.max_size,
+            max_queue=self.max_queue,
+            read_limit=self.read_limit,
+            write_limit=self.write_limit
+        )
+        websocket.subprotocol = subprotocol
+
+        # The protocol now opens the connection and gets pushed into place.
+        websocket.connection_made(request.transport)
+        websocket.connection_open()
+        request.transport.set_protocol(websocket)
+        return websocket
+
+    async def __call__(self, request, response, **params):
+        if request.upgrade != 'websocket':
+            # https://tools.ietf.org/html/rfc7231.html#page-62
+            # Upgrade needed but none was request or of the wrong type
+            response.status = HTTPStatus.UPGRADE_REQUIRED
+            response.body = b"This service requires the websocket protocol"
+            return
+
+        ws = self.switch_protocol(request)
+        if 'websockets' not in request.app:
+            request.app['websockets'] = set()
```

Comment:
How do we ensure that set is flushed after some time? Is that even pertinent? (I have no experience with websockets)

Reply:
The handler ensured that the websocket is removed from the set. In Sanic, to give an example, they keep all the connections in the same fashion but they clear it thanks to a task that verifies the timeouts. Here, I think I covered all in a try/catch. But there is a risk, if the global handled task is cancelled maybe.